pkg/api/store/nodetemplate/store.go

```diff
@@ -24,3 +33,61 @@ func (s *Store) Delete(apiContext *types.APIContext, schema *types.Schema, id st
 	}
 	return s.Store.Delete(apiContext, schema, id)
 }
+
+func (s *Store) Create(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}) (map[string]interface{}, error) {
+	if err := s.replaceCloudCredFields(data); err != nil {
+		return data, err
+	}
+	return s.Store.Create(apiContext, schema, data)
+}
+
+func (s *Store) Update(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}, id string) (map[string]interface{}, error) {
+	if err := s.replaceCloudCredFields(data); err != nil {
+		return data, err
+	}
+	return s.Store.Update(apiContext, schema, data, id)
+}
+
+func (s *Store) replaceCloudCredFields(data map[string]interface{}) error {
+	credID := convert.ToString(values.GetValueN(data, "cloudCredentialId"))
+	if credID == "" {
+		return nil
+	}
+	ns, name := ref.Parse(credID)
+	if ns == "" || name == "" {
+		return fmt.Errorf("invalid credID %s", credID)
+	}
+	cred, err := s.CloudCredentialLister.Get(namespace.GlobalNamespace, name)
+	if err != nil {
+		return fmt.Errorf("error getting cloud cred %s: %v", credID, err)
+	}
+	if len(cred.Data) == 0 {
+		return fmt.Errorf("empty credID data %s", credID)
+	}
+	configName, credConfigName := "", ""
+	for key := range cred.Data {
+		splitKey := strings.SplitN(key, "-", 2)
+		if len(splitKey) == 2 && strings.HasSuffix(splitKey[0], "credentialConfig") {
+			configName = strings.Replace(splitKey[0], "credential", "", 1)
+			credConfigName = splitKey[0]
+			break
+		}
+	}
+	if configName == "" {
+		return fmt.Errorf("empty configName for credID %s", configName)
+	}
+	toReplace := convert.ToMapInterface(values.GetValueN(data, configName))
+	if len(toReplace) == 0 {
+		return nil
+	}
+	var fields []string
```

Comment:
how this fields is being used? only for debug statement? 

Also can we iterate over cred.data only once? and do a replacement in one loop

Reply:
Yeah, just for debug, since we're emptying those fields - want to log them once, happens on only node template update 

the actual iteration is just once, earlier one breaks as soon as it gets the configName, will update to read from index 0